discojs/discojs-core/src/models/gpt/evaluate.ts

```diff
@@ -1,50 +1,58 @@
 import * as tf from '@tensorflow/tfjs'
 
+interface DataPoint extends tf.TensorContainerObject {
+  xs: tf.Tensor2D,
+  ys: tf.Tensor3D,
+}
+
 export default async function evaluate (
   model: tf.LayersModel,
-  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>
+  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>,
+  maxEvalBatches: number
 ): Promise<Record<'acc' | 'val_acc' | 'val_loss' | 'val_perplexity', number>> {
   let datasetSize = 0
   let totalLoss = 0
   const acc: [number, number] = [0, 0]
 
-  await dataset.map(({ xs, ys }) => {
+  let iteration = 1
+  const iterator = await dataset.iterator()
+  let continueIterating = true
+  while (continueIterating) {
+    const next = await iterator.next()
+
+    const { xs, ys } = next.value as DataPoint
```

Comment:
the iterator is not infinite, you will [read an invalid `next.value`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol) at the end of it (when `done` is true).
I don't really see the need of moving to a while loop instead of the stream/function interface: the latter takes care of the iteration for you, stopping when needed (use `dataset.take(maxEvalBatches)`).

Reply:
Yeah so I took the sad decision to use a while loop instead of the functional way because there was a memory leak when streaming over the dataset. I am still not exactly sure of where it happened but I suspect the iterator values were not correctly disposed. And a while loop allows me to ensure it is done.
The check for `next.done` is done at the end of the loop, but it assumes at least one datapoint in the dataset.

Reply:
Actually I thought you were talking about the training loop which I also converted from functional stream to an iterator loop and where the memory leak was occurring. 
There is still a growing number of allocated tensors (#654), especially in the evaluation loop, that's why I also tried changing the evaluate loop but it doesn't solve the issue. I'm reverting the evaluation loop to be functional.